### PR TITLE
[P2P] Improve peer selection part 3

### PR DESCRIPTION
**Pull request by fassadlr**
This is the last PR in the series.

**Peer selection and peer discovery now share the same selection logic.**

The basic outline is as follows:

```
First check to see if there are handshaked peers. If so,
give them a 75% chance to be picked over all the other peers.

If there are peers that have recently connected, give them
a 75% chance to be picked over fresh and/or attempted peers.

At this point, if neither selecting handshaked or connected
was successful, we will select from fresh or attempted.
If both sets exist, pick 50/50 between the two.

If there are only fresh peers, return them.

If there are only attempted peers, return them.

If all the selection criteria failed to return a set of peers,
then let the caller try again.
```
From the above set of peers we pick a random one.

### CODE REVIEW

**Comment by fassadlr** [comment]
@noescape00

**Comment by noescape00** [comment]
@fassadlr 
Can't test it since it always throws for me on NetworkPeerClient:94

```
-		$exception	{System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at System.Threading.ManualResetEventSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.SpinThenBlockingWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.InternalWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(CancellationToken cancellationToken)
   at Stratis.Bitcoin.P2P.Peer.NetworkPeerClient.<>c__DisplayClass21_0.<ConnectAsync>b__0() in C:\Users\user\Desktop\review\src\Stratis.Bitcoin\P2P\Peer\NetworkPeerClient.cs:line 94
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)}	System.OperationCanceledException

```

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs [hunk 1]
```diff
@@ -208,30 +208,35 @@ public void StartConnectAsync()
...
+                    timeoutTokenSource.CancelAfter(5000);
                     clonedConnectParamaters.ConnectCancellation = timeoutTokenSource.Token;
 
                     peer = await this.networkPeerFactory.CreateConnectedNetworkPeerAsync(this.network, peerAddress.NetworkAddress, clonedConnectParamaters).ConfigureAwait(false);
                     await peer.VersionHandshakeAsync(this.Requirements, timeoutTokenSource.Token).ConfigureAwait(false);
                 }
             }
+            catch (OperationCanceledException timeout)
+            {
+                this.logger.LogInformation("Peer {0} connection timeout.", peerAddress.NetworkAddress.Endpoint);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnector.cs]: i don't think this should be info level

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs [hunk 2]
```diff
@@ -208,30 +208,35 @@ public void StartConnectAsync()
...
+                peer?.DisconnectWithException("Timeout", timeout);
+
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnector.cs]: remove empty line

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs [hunk 3]
```diff
@@ -41,30 +47,98 @@ public sealed class PeerSelector : IPeerSelector
...
-            if (this.peerAddresses.Fresh().Any())
-                return this.peerAddresses.Fresh().Random();
+            PeerAddress peerAddress = null;
 
-            return null;
+            var peers = SelectPeers();
+            if (peers.Any())
+            {
+                peerAddress = peers.Random();
+                this.logger.LogTrace("(-):{0}={1}", nameof(peerAddress), peerAddress.NetworkAddress.Endpoint.ToString());
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerSelector.cs]: `this.logger.LogTrace("(-):'{0}'", peerAddress.NetworkAddress.Endpoint);`

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs [hunk 4]
```diff
@@ -41,30 +47,98 @@ public sealed class PeerSelector : IPeerSelector
...
+            }
+            else
+                this.logger.LogTrace("(-):{0} [NULL]", nameof(peerAddress));
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerSelector.cs]: `this.logger.LogTrace("(-)[NO_PEERS]");`

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs [hunk 5]
```diff
@@ -41,30 +47,98 @@ public sealed class PeerSelector : IPeerSelector
...
         public IEnumerable<PeerAddress> SelectPeers()
         {
-            return this.peerAddresses.Where(p => p.Value.Preferred).Select(p => p.Value);
+            this.logger.LogTrace("()");
+
+            // First check to see if there are handshaked peers. If so,
+            // give them a 75% chance to be picked over all the other peers.
+            if (this.peerAddresses.Handshaked().Any())
+            {
+                var chance = new Random().Next(100);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerSelector.cs]: don't create new random every time

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs [hunk 6]
```diff
@@ -41,30 +47,98 @@ public sealed class PeerSelector : IPeerSelector
...
+                    this.logger.LogTrace("(-)[RETURN_HANDSHAKED]");
+                    return this.peerAddresses.Handshaked();
+                }
+            }
+
+            // If there are peers that have recently connected, give them
+            // a 75% chance to be picked over fresh and/or attempted peers.
+            if (this.peerAddresses.Connected().Any())
+            {
+                var chance = new Random().Next(100);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerSelector.cs]: same here and below as well

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnector.cs]: OK this was where mentioned earlier in the slack channel, that it might be beneficial to the user to see that we are in fact trying to connect? We had this issue where the node was running for minutes without any connections, making it appear that it wasnt doing anything, where in actual fact it WAS trying to connect but just not being successful.

I don't really mind either way? WDYT?

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs (hunk 3 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerSelector.cs]: Fixed

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs (hunk 4 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerSelector.cs]: Fixed

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs (hunk 5 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerSelector.cs]: Fixed

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs (hunk 6 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerSelector.cs]: Fixed

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs (hunk 1 above)
Comment (Aprogiena) [reply src/Stratis.Bitcoin/P2P/PeerConnector.cs]: if you log it @ DEBUG level then using `-debug=1` without any additional settings, you would get this information being logged into `node.txt`, this seems to me simple enough for such scenarios to be checked

Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnector.cs]: Ok cool!

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs [hunk 7]
```diff
@@ -130,22 +206,14 @@ public static IEnumerable<PeerAddress> Handshaked(this ConcurrentDictionary<IPEn
...
-
-                if (random.Next(1 << 30) < chanceFactor * randomPeer.Selectability * (1 << 30))
-                    return randomPeer;
+            if (peers.Count() == 1)
+                return peers.First();
 
-                chanceFactor *= 1.2;
-            }
+            var randomPeerIndex = random.Next(peers.Count() - 1);
+            var randomPeer = peers.ToArray()[randomPeerIndex];
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerSelector.cs]: ElementAt would be better imho

**Comment by Aprogiena** [comment]
btw this
```
        /// <para>
        /// Refer to <see cref="IPeerSelector.SelectPeer()"/> for more details.
        /// </para>
```
can be written in shorter way (outside summary section) as
```
        /// <seealso cref="IPeerSelector.SelectPeer()"/>
```

**Comment by noescape00** [comment]
@fassadlr about that bug when VS would throw handled exceptions: there is a workaround instead of disabling "Just my code" option- you can put whole method inside the task like this: 

 ```
 public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation)
        {
            this.logger = this.loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.Id}-{endPoint}] ");
            this.logger.LogTrace("({0}:'{1}')", nameof(endPoint), endPoint);

            await Task.Run(() =>
            {
                try
                {
                    this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).Wait(cancellation);
                    this.Stream = this.tcpClient.GetStream();
                }
                catch (OperationCanceledException)
                {
                    this.logger.LogTrace("Connecting to '{0}' cancelled.", endPoint);
                    this.logger.LogTrace("(-)[CANCELLED]");
                    //throw;
                }
                catch (Exception e)
                {
                    if (e is AggregateException) e = e.InnerException;
                    this.logger.LogDebug("Error connecting to '{0}', exception: {1}", endPoint, e.ToString());
                    this.logger.LogTrace("(-)[UNHANDLED_EXCEPTION]");
                    //throw e;
                }
                this.logger.LogTrace("(-)");
            }).ConfigureAwait(false);
        }
```

It looks like VS thinks that if exception was not handled inside the task it was unhandle so the debugger should stop. 

Also you can do the same with AcceptClientsAsync() since it also throws sometimes on `acceptTask.Wait(this.serverCancel.Token);` - "System.OperationCanceledException: 'The operation was canceled.'"

**Comment by noescape00** [comment]
I think you should replace > with < because it selects peers that satisfied the condition anytime before (now - 1 minute). Same with `Attempted `and `Handshaked`. So effectively we are selecting peers that we've connected to anytime before.

```
        /// <summary>
        /// Return peers which've had successful connection attempts.
        /// <para>
        /// The result filters out peers which satisfies the above condition within the
        /// last 60 seconds.
        /// </para>
        /// </summary>
        public static IEnumerable<PeerAddress> Connected(this ConcurrentDictionary<IPEndPoint, PeerAddress> peers)
        {
            var result = peers.Skip(0).Where(p =>
                                p.Value.Connected &&
                                p.Value.LastConnectionSuccess < DateTime.UtcNow.AddSeconds(-60)).Select(p => p.Value);
            return result;
        }
```

Also peers from peerAddresses.Handshaked() and peerAddresses.Connected() are **the same peers** usually.


So imagine if we have ever connected to at least 8 different nodes- we'd have peerAddresses.Connected().Count == 8 and the max connections count is 8. So in this case when we do call SelectPeers() we have a 93.75% chance of getting a list of nodes that we've already connected anytime before. (probability of not getting Handshaked list is 0.25, probability of not getting Connected list is also 0.25. Probability of those 2 actions in the same time is 0.25*0.25 = 0.0625 = 6.25%. And since probability of not getting is 6.25 => probability of getting is 100-6.25 = 93.75%)

Probability of getting this list 8 times in a row is 0.9375^8 = 0.59671947383 == 59%. 

So with the current algorithm we will reconnect to the old nodes **without even trying a single fresh one** is **59%**.

**Comment by noescape00** [comment]
On the other hand if you fix the logic on the node's start the "Connected" list would be empty so we will be reduced to selecting randomly from the fresh and attempted. And at this point old algorithm is better because it was mostly selecting from the ones that have higher Selectability value.

**Comment by fassadlr** [comment]
@noescape00 thanks for your input!

> I think you should replace > with < because it selects peers that satisfied the condition anytime before (now - 1 minute). Same with Attemptedand Handshaked. So effectively we are selecting peers that we've connected to anytime before.

This is by design as we want to "throttle" the list from returning a peer we **just** connected to (or handshaked). Yes, connected peers **are** usually handshaked peers but what if we, in the future, start releasing new versions of FullNode more often, then we can possibly sit with a situation where there a equal amount of peers that could connect as well as a set of that we could handshake with. By splitting the list now already we future proof that scenario.

> So with the current algorithm we will reconnect to the old nodes without even trying a single fresh one with probability of 59%.

This to me seems acceptable, as we always want to first try and connect to "good" peers, and leaving the fresh or attempted ones at the back.

**However,** we can alter that 75% chance down to say 60%, so that we give the fresh peers a bigger chance to be selected up on startup?

Also, @noescape00, @Aprogiena and I chatted and I realized that peer connection and peer discovery/getAddr payload cant share the same algo. So the distinction also has to be made there.

Im going to commit some changes to these two different sets of logic.

**Comment by noescape00** [comment]
>This is by design as we want to "throttle" the list from returning a peer we just connected to (or handshaked).

Then I'd suggest changing the comments for those methods. Right now this comment: 
`The result filters out peers which satisfies the above condition within the last 60 seconds.`  seems misleading. 

Probably change it to be like that: 
`The result filters out peers which satisfies the above condition (excluding the peers that started to satisfy it during the last 60 seconds).`

**Comment by noescape00** [comment]
>Yes, connected peers are usually handshaked peers but what if we, in the future, start releasing new versions of FullNode more often, then we can possibly sit with a situation where there a equal amount of peers that could connect as well as a set of that we could handshake with. By splitting the list now already we future proof that scenario.

Makes sense, agreed. 


>>So with the current algorithm we will reconnect to the old nodes without even trying a single fresh one with probability of 59%.

>This to me seems acceptable, as we always want to first try and connect to "good" peers, and leaving the fresh or attempted ones at the back.

The problem is that we will always connect to the same nodes we've discovered: 

Imagine if we have ever discovered, lets say, 20 nodes. They are in the connected and handshaked list. 
When we launch a node most likely we will connect to 8/20 of those nodes. During the run we will eventually disconnect from time to time from some of those 8 but we will connect to some nodes from that list again. At some point we will hardly ever connect to a new node.

**Comment by fassadlr** [comment]
> The problem is that we will always connect to the same nodes we've discovered:

Ah but I think that you have missed the fact that we select a **random**  peer from whatever list was selected ;)

```
    var peers = SelectPreferredPeers();
    if (peers.Any())
    {
        peerAddress = peers.Random();
        this.logger.LogTrace("(-):'{0}'", peerAddress.NetworkAddress.Endpoint);
    }
```

**Comment by fassadlr** [comment]
> Imagine if we have ever discovered, lets say, 20 nodes. They are in the connected and handshaked list.
> When we launch a node most likely we will connect to 8/20 of those nodes

That is why I said that we can possibly look at changing that chance factor to 60%. I.e. so that we gice connected/handshaked peers 60% to be selected over fresh/attempted?

**Comment by noescape00** [comment]
Lets assume that we will be able to connect to 1 of 50 tried fresh nodes. 
Also let's assume that on average after we've connected to the peer we will be connected to it on average for 10 minutes.

Chance of selecting a peer NOT from connected\handshaked list is 6.25%

Chance of selecting a fresh address AND connecting to it is 0.0625 * (1/50) = 0.00125 == 0.125%

Chance of this not happening is 100 - 0.125 = 99.875%

And we're rolling this dice every 10/8 minutes (as soon as 1 of 8 max nodes is disconnected we need a new connection)
10/8 = 1.25 minutes - delay between rolling that dice. 

Chance of not getting a single connection from fresh peers in 1 hour is: 
0.99875^(1*60/1.25) = 0.94 == 94%

Chance of not getting a single connection from fresh peers in 11 hour is: 
0.99875^(11*60/1.25) = 0.5166 == 51%

Chance of not getting a single connection from fresh peers in 24 hours is: 
0.99875^(24*60/1.25) = 0.2367 == 24%

Chance of not getting a single connection from fresh peers in a week  is: 
0.99875^(7*24*60/1.25) = 0.004164 == 0%

Bottom line is: 
**We have a REALLY high chance of not connecting to a new peer for a LONG time.** 
We have a 50% chance that we get a fresh connection if we wait for ~11 hours.

**Comment by noescape00** [comment]
With that factor 60% instead of 75%: 

X = 1 - 0.4*0.4 * (1/50) = 0.9968

Chance of not getting a single connection from fresh peers in 1 hour is:
0.9968^(1*60/1.25) = 0.857 = **85%** 

Chance of not getting a single connection from fresh peers in 24 hours is:
0.9968^(24*60/1.25) = 0.02 = **2%** 

Well, this is better but I think chances are still very high.

**Comment by noescape00** [comment]
>>The problem is that we will always connect to the same nodes we've discovered:

>Ah but I think that you have missed the fact that we select a random peer from whatever list was selected ;)

No, I've noticed that.

Sorry, I probably was unclear in that message. What I wanted to say that we will be always connected to a subset of good nodes we've discovered.

**Comment by noescape00** [comment]
I'd suggest using this algorithm for getting first 1-2 connections and an old one for the rest cases. 
This will both ensure getting first few connections when the node started fast enough and will be random enough so we can add new connections and update this Connected list frequently enough. 

I'd also suggest limiting the connected and handshaked list by 100-500 most recently updated ones.

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs [hunk 8]
```diff
@@ -41,30 +43,241 @@ public sealed class PeerSelector : IPeerSelector
...
+
+            // If all the selection criteria failed to return a set of peers,
+            // then let the caller try again.
+            this.logger.LogTrace("(-)[RETURN_NO_PEERS]");
+            return new PeerAddress[] { };
         }
 
         /// <inheritdoc/>
-        public IEnumerable<PeerAddress> SelectPeers()
+        public IEnumerable<PeerAddress> SelectPeers(int amountOfPeers)
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerSelector.cs]: peerCount

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs [hunk 9]
```diff
@@ -41,30 +43,241 @@ public sealed class PeerSelector : IPeerSelector
...
+            // If there is only 1 set, take the amount of peers requested.
+            if (handshaked.Any() && !connected.Any() && !attemptedAndFresh.Any())
+            {
+                return handshaked.Take(amountOfPeers);
+            }
+
+            // If there is only 1 set, take the amount of peers requested.
+            if (!handshaked.Any() && connected.Any() && !attemptedAndFresh.Any())
+            {
+                return connected.Take(amountOfPeers);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerSelector.cs]: this if series is hardly readable, can we improve it

Comment (dangershony) [reply src/Stratis.Bitcoin/P2P/PeerSelector.cs]: yes I agree it could be beteter

Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerSelector.cs]: @dangershony This is an old commit ;) This logic has been removed 👍

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs [hunk 10]
```diff
@@ -240,7 +239,7 @@ public void StartReceiveMessages()
         {
             this.logger.LogTrace("()");
 
-            this.receiveMessageTask = ReceiveMessagesAsync();
+            this.receiveMessageTask = this.ReceiveMessagesAsync();
```
Comment (dangershony) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: such a big pr you should try to contain changes to only your logic ;)

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs [hunk 11]
```diff
@@ -16,55 +16,144 @@ public interface IPeerSelector
...
+
+            // First check to see if there are handshaked peers. If so,
+            // give them a 50% chance to be picked over all the other peers.
+            if (this.peerAddresses.Handshaked().Any())
+            {
+                int chance = this.random.Next(100);
+                if (chance <= 50)
+                {
+                    this.logger.LogTrace("(-)[RETURN_HANDSHAKED]");
+                    return this.peerAddresses.Handshaked();
```
Comment (dangershony) [inline src/Stratis.Bitcoin/P2P/PeerSelector.cs]: Is there any randomness in the HandShaked() method?

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs [hunk 12]
```diff
@@ -16,55 +16,144 @@ public interface IPeerSelector
...
+                }
+                else
+                {
+                    this.logger.LogTrace("(-)[RETURN_FRESH]");
+                    return this.peerAddresses.Fresh();
+                }
+            }
+
+            // If there are only fresh peers, return them.
+            if (this.peerAddresses.Fresh().Any() && !this.peerAddresses.Attempted().Any())
```
Comment (dangershony) [inline src/Stratis.Bitcoin/P2P/PeerSelector.cs]: Multiple enumerations cant we find a way to reduce the enumeration count?

File: src/Stratis.Bitcoin/P2P/PeerSelector.cs (hunk 11 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerSelector.cs]: No, a random peer is picked in `SelectPeer` from whatever set of peers is returned in `SelectPreferredPeers`.

**Comment by dangershony** [comment]
The algo looks good to me, defo better then before.
I think @noescape00 has some good comments as well so perhaps we should open an issue on how we could improve on this algo further.
For now I am happy with pushing this as is.

**Comment by noescape00** [comment]
@dangershony I'm fine with pushing it. 
All in all this algorithm is better in terms of getting first few connections way faster then the old one. 
Also I've discovered that it performs much better then the old one if the target is to get as many connections as possible (setting max connections to >1000).

So this should do for now and we can improve randomness of this algorithm in the future.
